sdk/common/src/main/java/io/opentelemetry/sdk/common/export/MemoryMode.java

```diff
@@ -17,6 +17,21 @@ public enum MemoryMode {
    *
    * <p>In this mode, the SDK reuses objects to reduce allocations, at the expense of disallowing
    * concurrent collections / exports.
+   *
+   * <p>Metric Signal: For DELTA aggregation temporality, the memory used for recording and
+   * aggregating Attributes metric values is kept between MetricReader collect operation, to avoid
+   * memory allocations. In the case of DELTA, when the configured maximum cardinality of Attributes
+   * is reached, unused Attributes are cleared from this memory, during collect operation, at the
+   * cost of memory allocations that will be done next time those attributes are used. Those
+   * allocations can be minimized by increasing the configured max cardinality. For example, a user
+   * has recorded values for 1000 unique Attributes, while the max cardinality configured was 2000.
+   * If after a collections only 100 unique Attributes values are recorded, the MetricReader's
+   * collect operation would return 100 points, while in memory the Attributes data structure keeps
+   * 1000 unique Attributes. If a user recorded values for 3000 unique attributes, the values for
+   * the first 1999 Attributes would be recorded, and the rest of 1001 unique Attributes values
+   * would be recorded in the CARDINALITY_OVERFLOW Attributes. If after several collect operations,
+   * the user now records values to only 500 unique attributes, during collect operation, the unused
+   * 1500 Attributes memory would be cleared from memory.
```

Comment:
Want to call out this paragraph to other @open-telemetry/java-approvers / @open-telemetry/java-maintainers - @asafm and I have talked about this, but others should be aware of / comment on this behavior. 

Essentially, when temporality is DELTA and memory mode is REUSEABLE, you might have a cardinality limit of 2000 for an instrument but only get back 1000 series. This is the result of REUSEABLE memory mode prioritizing low allocations, which is implemented by holding onto the aggregators for series even when they don't have measurements. The behavior of IMMUTABLE mode with delta is to remove the aggregators from the ConcurrentHashMap and put them in an object pool for reuse, but re-adding the entries to the map causing memory allocations.

Reply:
(I haven't looked at any of the code, so could be way off base)

would it work to set a boolean field on the reusable objects back to false after collect, and then keeping a running count of how many reusable objects flipped back to true during the interval, so we can "degrade" to creating new objects right away if needed, for "correctness"

Reply:
@trask I believe (correct if I'm wrong) @jack-berg meant a different problem. Say before the 1st collection you recorded to attributes A1-A1000. Same before 2nd collection. 
Then before 3rd collection, you recorded only to attributes A1001-A2000. In this case, the response you will get in the 3rd collection is a list of 1000 points (for this instrument) + 1 point of the OVERFLOW Attributes. Also you will get a warning in the log saying you crossed the amount of unique attributes allowed.  
The user might confused: He set cardinality limit of 2000, got 1000 points in the last collection and this warning about the limit crossed, but 1000 < 2000 (limit). Before collection 5th, the map would already be clear of attributes A1-A1000, and only filled with Attributes A1001-A1999 (with value = 0).

This only happens in MemoryMode.RESUABLE_DATA which has to be set explicitly, thus you are supposed to read the javaDoc of MemoryMode which explains this.

What you are suggesting is to try to optimize and clear the map of unused handlers during recording, and not wait for collect(). We do have the flag you suggest, on the handlers (`hasRecordedValues`). The only issue is that we'll have to keep another data structure to know which of the elements in the handlers map are free to evacuate (`hasRecordedValues`=false). Otherwise we'll have O(N) search to find in the map values which has this flag as false. This will be done per recording() which is the hot path. Not sure it's a good thing.